base/est/src/main/java/org/dogtagpki/est/ESTRequestAuthorizer.java

```diff
@@ -52,4 +64,123 @@ public abstract Object authorizeSimplereenroll(
+        } else {
+            // construct X509CertImpl
+            try {
+                cert = new X509CertImpl(cert_.getEncoded());
+            } catch (CertificateException e) {
+                throw new ForbiddenException("Failed to decode user certificate.");
+            }
+        }
+
+        // Compare Subject DNs.
+        //
+        // This comparison does not perform StringPrep or caseIgnoreMatch.
+        // However, RFC 7030 says the values must be "identical", not "equal"
+        // or "equivalent", so this seems reasonable.
+        //
+        // In case of new certificate the check is less strict and done after string conversion
+        //
+        if (!csr.getSubjectName().equals(cert.getSubjectName()) &&
+                (!renew && !csr.getSubjectName().toString().equals(cert.getSubjectName().toString()))) {
+            throw new ForbiddenException("CSR subject does not match user certificate.");
+        }
+
+        // Compare SAN
+        SubjectAlternativeNameExtension csrSAN = null;
+        try {
+            csrSAN = (SubjectAlternativeNameExtension)
+                CryptoUtil.getExtensionFromPKCS10(csr, SubjectAlternativeNameExtension.NAME);
+        } catch (IOException | CertificateException e) {
+            throw new BadRequestException("Failed to decode SAN extension in CSR.");
+        }
+
+        // TODO get SAN from t-b-r cert; compare
+        SubjectAlternativeNameExtension certSAN = (SubjectAlternativeNameExtension)
+            cert.getExtension(PKIXExtensions.SubjectAlternativeName_Id.toString());
+
+        // In case of new certificate the SAN can be missed in the CSR and
+        // if present the controls are done on Strings
+        if (csrSAN != null && certSAN != null) {
+            if (!Arrays.equals(csrSAN.getExtensionValue(), certSAN.getExtensionValue()) &&
+                    (!renew && !csrSAN.toString().equals(certSAN.toString()))) {
+                throw new ForbiddenException(
+                    "SAN extensions of user certificate and CSR are not identical.");
+            }
+        } else if (csrSAN == null && certSAN != null && renew) {
+            throw new ForbiddenException(
+                "User certificate has SubjectAlternativeName extension, "
+                + "but CSR does not."
+            );
+        } else if (csrSAN != null && certSAN == null) {
+            throw new ForbiddenException(
+                "User certificate does not have SubjectAlternativeName extension, "
+                + "but CSR does."
+            );
+        } // else both null, which is valid
+    }
+
+   /** Ensure subject info in CSR matches the user details.
+    *
+    * This function implements that requirement.
+    *
+    * @throws ForbiddenException if fields are not identical.
+    */
+   protected static void ensureCSRMatchesToBeCert(PKCS10 csr, User user)
+           throws ForbiddenException {
+
+
+       String subject = csr.getSubjectName().toString().substring(3);
```

Comment:
This code assumes that the subject name uses a 2-character attribute name such as `CN` so it won't work with `UID` attribute. Should we use a proper DN parsing code? Should we check what attribute it's actually using?